### PR TITLE
Fix: replace is_connect_failed with is_err

### PR DESCRIPTION
**Pull request by irvingoujAtDevolution**
In linux, epoll, EPOLLHUP may happen even if no connection call is made. It would confuse callers for what is actually happening.
Replaced is_connect_failed, and we detect if connection failed by using the combination of is_err and is_interrupt, please see the example, tcp_client

### CODE REVIEW

**Review by notgull — CHANGES_REQUESTED** [review]
Unfortunately we can't change `is_connect_failed` as that would be a breaking change. So we have to keep the name the same there.

**Comment by irvingoujAtDevolution** [comment]
> Unfortunately we can't change `is_connect_failed` as that would be a breaking change. So we have to keep the name the same there.

I apologize for my previous mistake. I would like to have your opinion on whether it is better to add `is_err`, mark `is_connect_fail` as deprecated, or simply keep using the same name but change the implementation.

I would argue that the first approach is better, as it aligns with the semantics it provides.

**Comment by notgull** [comment]
> I apologize for my previous mistake. I would like to have your opinion on whether it is better to add `is_err`, mark `is_connect_fail` as deprecated, or simply keep using the same name but change the implementation.
> 
> I would argue that the first approach is better, as it aligns with the semantics it provides.

I'd argue that the first would be better.

File: examples/tcp_client.rs [hunk 1]
```diff
@@ -1,36 +1,155 @@
-use std::{io, net};
```
Comment (notgull) [inline examples/tcp_client.rs:1]: I would appreciate not changing this example and instead just wrapping the socket error bits in the configuration guards.

File: examples/tcp_client.rs [hunk 2]
```diff
@@ -1,36 +1,155 @@
...
+
+    assert!(event.writable);
+    assert!(!event.is_interrupt());
+    assert!(!event.is_err().unwrap());
+
+    println!("event: {:?}", event);
+    println!("socket is now writable");
+    // ========================================================================
+    // the below is example of a bad socket
+    println!("testing bad socket");
```
Comment (notgull) [inline examples/tcp_client.rs:55]: If this is a functionality test it should be moved to be under the `tests` folder.

File: src/lib.rs [hunk 3]
```diff
@@ -391,10 +391,63 @@ impl Event {
     /// Returns `Some(true)` if the connection has failed, `Some(false)` if the connection has not failed,
     /// or `None` if the platform does not support detecting this condition.
     #[inline]
+    #[deprecated]
```
Comment (notgull) [inline src/lib.rs:394]: Please add a deprecation message here.

**Review by notgull — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 3 above)
Comment (irvingoujAtDevolution) [reply src/lib.rs]: updated

File: examples/tcp_client.rs (hunk 1 above)
Comment (notgull) [reply examples/tcp_client.rs]: Don't modify the Example too much; just make it so the `is_err`-specific code is guarded.

**Comment by irvingoujAtDevolution** [comment]
please have a look again, I am not exactly sure if this is how you want it.

File: src/lib.rs [hunk 4]
```diff
@@ -391,10 +391,66 @@ impl Event {
...
     pub fn is_connect_failed(&self) -> Option<bool> {
         self.extra.is_connect_failed()
     }
 
+    /// Tells if this event is the result of a connection failure.
+    ///
+    /// This function checks if an error exist,particularlly useful in detecting if TCP connection failed. It corresponds to the `EPOLLERR` event in Linux
+    /// and `CONNECT_FAILED` event in Windows IOCP.
+    ///
+    /// note: in epoll, a tcp connection failure is indicated by EPOLLERR + EPOLLHUP, though just EPOLLERR is enough to indicate a connection failure.
```
Comment (notgull) [inline src/lib.rs:407]:
```suggestion
    /// ## Caveats
    ///
    /// In `epoll`, a TCP connection failure is indicated by `EPOLLERR` + `EPOLLHUP`, though just `EPOLLERR` is enough to indicate a connection failure.
```

File: src/lib.rs [hunk 5]
```diff
@@ -391,10 +391,66 @@ impl Event {
...
+    ///           None => {
+    ///               println!("not supported in this platform");
+    ///               return Ok(());
+    ///           },
+    ///       };
+    ///       assert!(is_err);
+    ///       println!("bad socket is now in error state");
+    ///   
+    ///       Ok(())
+    ///   }
```
Comment (notgull) [inline src/lib.rs:445]: Why is this block indented?

File: src/lib.rs [hunk 6]
```diff
@@ -391,10 +391,66 @@ impl Event {
...
+    /// ```  
+    /// Returns `Some(true)` if the connection has failed, `Some(false)` if there is an error,
```
Comment (notgull) [inline src/lib.rs:447]:
```suggestion
    /// ```  
    ///
    /// Returns `Some(true)` if the connection has failed, `Some(false)` if there is an error,
```

**Review by notgull — CHANGES_REQUESTED** [review]
Overall LGTM, just some minor considerations regarding docs.

File: src/lib.rs (hunk 5 above)
Comment (irvingoujAtDevolution) [reply src/lib.rs]: updated, removed it

File: src/lib.rs [hunk 7]
```diff
@@ -391,10 +391,31 @@ impl Event {
     /// Returns `Some(true)` if the connection has failed, `Some(false)` if the connection has not failed,
     /// or `None` if the platform does not support detecting this condition.
     #[inline]
+    #[deprecated(
+        since = "3.4.0",
+        note = "use `is_err` in combination of is_hub instead, see documentation for `is_err`"
```
Comment (notgull) [inline src/lib.rs:396]:
```suggestion
        note = "use `is_err` in combination of is_hup instead, see documentation for `is_err`"
```

**Review by notgull — COMMENTED** [review]
LGTM aside from one minor nitpick

**Comment by irvingoujAtDevolution** [comment]
😢 , @notgull could you please help me look at the wine CI thing? I have no idea how that works

**Comment by notgull** [comment]
Looks like a [spurious error](https://github.com/smol-rs/polling/actions/runs/8363127534/job/22895359605).

**Review by notgull — APPROVED** [review]
Thanks!

**Pull request merged**